### DISCUSSION TITLE
HS_NOTCH

### DISCUSSION THREAD

**Discussion by ja2ui0** [joshreve/dactyl-keyboard#15] [General]
@joshreve mentioned in the config generator code that HS_[NOTCH|UNDERCUT] does not generate properly and the hot swap step needs to be modified.

I generated a HS_NOTCH case this afternoon and I want to make sure I'm seeing the same thing before I go down this rabbit hole. To the untrained (possibly including my own) eye, it looks pretty good. Though on closer inspection the only thing that jumps out is that the sides of the hot swap tray come in closer to the key body than the depth of the notch. Is this the part that needs to be modified, so that the notch depth is the important bit and the hot swap tray doesn't interfere with that?

I have not handwired one of these yet, so I can only trust that the orientation of the sockets is fine (particularly around where the thumb cluster is concerned). I'm sure it's a tight / awkward fit with an iron no matter what when you get near the case edges, but let's not get into the particulars of rotating switches at this juncture :smile: - just so long as none of the rotations are absolute garbage from a wiring perspective.

Assuming all of that is in order, I can work on that step file and submit a PR shortly, just let me know if the bit I mentioned is the part of concern or if there's something I'm just completely missing.

**Comment by joshreve** [comment]
That was the challenge.  Originally it had a third wall blocking the mount notch near the hotswap.  I actually half fixed them after that comment by removing the third wall.  I'm just not sure how strong it is without the added structure.  It should theoretically work, but I'm really not sure how it will print.  If it prints and works well, I'd love to remove the warning, but at the moment it is more of a "not sure" rather than a "guaranteed to fail".  I added rotations in the cluster to orient them if you have issues, but that is a WIP setting.  

You may want to print just a thumb cluster as a trial to check the fit in that area and the notch settings as well.

**Reply by ja2ui0** [reply]
I'll do that. I have a feeling that the rotation for 1u default thumb cluster makes the hot swap mount weird on tl (right side), requiring a rotation, which is only important from a wiring standpoint, but that's hard for me to determine not having done it yet. 

As for the notch situation, I'm not sure what you mean by third wall. If the current situation is "not sure" as you say then I'll try a test print tomorrow, but I won't be able to actually fit anything until my sockets show up next week. Without those physically in hand, I'm not sure what to be looking for with regards to stability but I'm open to suggestions. 

It seems like if switches are held fast by the notches (they are without hs, anyway) then there should be no change in that by the addition of the hot swap mounts. Is the stability concern for that of the sockets themselves?

**Reply by ja2ui0** [reply]
I spent most of the night inspecting and working on the model for the hs socket holder, here's what I've found.

To start with, I think your removing the rear wall was not only fine to simplify the socket model, but the wall provided zero additional stability to the assembled model. I suggest this due to the fact that the rear of the holder is connected to the bottom of the key hole anyway, so the rear wall was overlapping / coincident plastic.

I actually got to the point where I'd carved away the entirety of the right wall (viewed from the front) with the same condition as the rear wall, the top of the holder still connected to the bottom of the key hole, so removing the wall should make no difference other than increasing the depth of the notch to whatever the user sets it to. No loss of stability I would think for the same reason as the rear one.

The left side is only slightly trickier because there's what looks like a diode holder there, and an angle going up toward the key hole perimeter. I don't think it would be a problem to cut that off so the entire top surface of the holder where the key sits is flush, but I hesitated to do that because I wanted the angled part to have a solid connection to the key hole perimeter. Regardless of whether the whole side is trimmed off or not, it is totally possible to take away at least the inner surface of the left side of the holder .5mm deeper to be flush with the default notch depth with no loss of stability from a purely observational standpoint.

"So ja2ui0, why don't you submit a PR then?"

I got the model all carved up over a couple of hours, and I can't figure out why every time I export a step file, it's in the wrong orientation when the model builds and renders. After playing with the rotation in freecad for a while, I did get what looked like the proper rotation and placement, then I noticed that for some reason, a solid surface (might have been a zero thickness plane, but it was visible in the model) was drawn at the bottom of the center hole where the switch inserts. So I flipped the table over and wrote this comment instead.

Right now I'm printing a thumb cluster with the original step file and the 1.5u setting, because whatever you did in the most recent version of the repo I pulled last night seems to have fixed the key overlap for the 1.5u keybed in the tl location as well. That wasn't indicated and seemed to be an issue with the original clojure-driven geometry as well, but I thought it was pretty cool and decided it was worth double checking. While I'm mentioning it, the 1u tr key definitely has a weird rotation; -90 degrees makes it identical rotation to the 1.5u. This might matter for wire placement.

So I can have another go at the step file this week, but if you've already resolved the rotation issue or don't encounter the same thing in practice, the comments above about wall removal should be reliable.

**Reply by ja2ui0** [reply]
My test print is complete and I tried a switch fit. The stability looks totally fine, and the notches seem fine. The problem I think I'm running into is purely print related and not an issue with the design - printing at .2mm resolution produces HS mounts that are barely passable in terms of structure and tolerance particularly around the holes the pins go through, and I think I'm getting parts that are only barely slightly unlevel around those holes causing my switches to be high on the bottom side, only flush on the top side. It looks like this is due to plastic on the bottom side of the pin holes nudging up the bottom of the switch.

If that's going to be the case, again not a design issue but a fabrication one, I have to decide if it's worth the outcome for me to awkwardly hand sand every single key hole, or redesign the mount, or give up on hotswap altogether. Not sure where I'm at with it yet. I will probably take a swing at redesigning the mount entirely so it's a little less articulated and accommodating to .2mm resolution prints, or back off the idea entirely. Or buy a resin printer.

**Reply by joshreve** [reply]
Yeah, those rotations are a mess.  I have to look at them again.  They were working in most cases, but I don't have all of them straight in my head.  The FreeCAD export is a pain to get the right orientation and that isn't my file originally.  It is a mess that I haven't attacked yet since it seemed sufficient for the moment.  

For the hotswap file, if you want to do a PR with the FreeCAD file as a different name I can pull it in to look at and either modify the current one or add a HS variant parameter and do the exports.  I feel like there may be others that show up along the way like a choc hotswap or something like that.  The FreeCAD geometry creation is the harder part, I'm practiced at this point beating the exports into submission. 

On a positive note, it seems to be generally working with just a few glitches that I should be able to fix.  Can you give a short bullet list of what you see as an issue?  You've given great info, but I'm trying separate what might be a little odd from what is definitely a problem.

**Reply by joshreve** [reply]
FYI, just checked the poll and minidox is beating everything so I will likely throw it in the mix.  Probably will get to it next weekend.

**Reply by Looooopy** [reply]
I tried to generate with HS_NOTCH but got 0.2 mm wall thickness on the thumb cluster **highlighted part**  which is to thin I think and it do not look the same as the standard cluster of button. 

![image](https://user-images.githubusercontent.com/16755871/143153752-3065c379-1747-4558-9cb2-a5b3a9cf4dea.png)

I did see the comment, but I don't know it is that that give me the incorrect output.
>HS_NOTCH' = hot swap underside with notch.  Does not generate properly.  Hot swap step needs to be modified.

Any idea if it should be like that or is it an issue with the thumb cluster and HS_NOTCH?

**Reply by Looooopy** [reply]
Seams like there are a problem in the source code, I commented out this line after some digging:
https://github.com/joshreve/dactyl-keyboard/blob/master/src/dactyl_manuform.py#L949

And now they are back, but I don't know if I screwed up something else :)
@joshreve How about it, did I do it correctly?

![image](https://user-images.githubusercontent.com/16755871/143500702-209204eb-a7d0-4302-aa48-210deade3bea.png)

generate_configuration.py changes I made
```python
...
'ncols':  6, #6,  # key columns
...
'default_1U_cluster': False
...
#################
    ## Switch Hole ##
    #################
'plate_style': 'HS_NUB',

'oled_configurations': {
...
    CLIP': {
            # Common parameters
            'oled_mount_width': 27.0,  # whole OLED width
            ...
            'oled_left_wall_x_offset_override':  38.5,
            ....
            'oled_mount_connector_hole': 5.0,
            'oled_screen_start_from_conn_end': 5.0,
            'oled_screen_length': 15.5,
            'oled_screen_width': 25.0,
            ...
    }
}
...
'controller_mount_type':  'RJ9_USB_WALL',
```

**Reply by joshreve** [reply]
There is a setting to clear space behind the socket to make sure you can mount a pcb.  It looks like either it is set on the cluster only or it is a bug and the seeing isn't applied consistently.  I will look.  I'm guessing that is why it is present but trimmed back in the cluster.

**Reply by joshreve** [reply]
Double check this is set  
 ​    ​'plate_pcb_clear'​: ​False​,
I'm guessing it is and I'm clearing in the cluster anyway due to a bug.  I'll take a look.

**Reply by joshreve** [reply]
Checking the line, you are fine to comment that out.  It should be within a conditional based on the parameter in my previous response.  I will fix it moving forward.  In the meantime your change shouldn't cause any issues since it is exactly what I thought it was and is not used far any other purpose.

**Reply by Looooopy** [reply]
Thanks for that

**Reply by joshreve** [reply]
Threw the bugfix in.  Should be commented out in the master branch now.
